components/layout_2020/table/layout.rs

```diff
@@ -918,88 +1007,356 @@ impl<'a> TableLayout<'a> {
+                current_rows_size -
+                border_spacing_spanned)
+                .max(Au::zero());
+
+            self.distribute_extra_size_to_rows(
+                excess_size,
+                rows_spanned,
+                row_sizes,
+                None,
+                true, /* rowspan_distribution */
+            );
+        }
+    }
+
+    /// An implementation of the same extra block size distribution algorithm used in
+    /// LayoutNG and described at <https://github.com/w3c/csswg-drafts/issues/4418>.
+    fn distribute_extra_size_to_rows(
+        &self,
+        mut excess_size: Au,
+        track_range: Range<usize>,
+        track_sizes: &mut Vec<Au>,
+        percentage_resolution_size: Option<Au>,
+        rowspan_distribution: bool,
+    ) {
+        if excess_size.is_zero() {
+            return;
+        }
+
+        let is_constrained = |track_index: &usize| self.rows[*track_index].constrained;
+        let is_unconstrained = |track_index: &usize| !is_constrained(track_index);
+        let is_empty: Vec<bool> = track_sizes.iter().map(|size| size.is_zero()).collect();
+        let is_not_empty = |track_index: &usize| !is_empty[*track_index];
+        let other_row_that_starts_a_rowspan = |track_index: &usize| {
+            *track_index != track_range.start &&
+                self.rows[*track_index].has_cell_with_span_greater_than_one
+        };
+
+        // If we have a table height (not during rowspan distribution), first distribute to rows
+        // that have percentage sizes proportionally to the size missing to reach the percentage
+        // of table height required.
+        if let Some(percentage_resolution_size) = percentage_resolution_size {
+            let get_percent_block_size_deficit = |row_index: usize, track_size: Au| {
+                let size_needed_for_percent = percentage_resolution_size
+                    .scale_by(self.rows[row_index].percent.0 as f32 / 100.);
+                size_needed_for_percent - track_size.max(Au::zero())
+            };
+            let percent_block_size_deficit: Au = track_range
+                .clone()
+                .map(|index| get_percent_block_size_deficit(index, track_sizes[index]))
+                .sum();
+            let percent_distributable_block_size = percent_block_size_deficit.min(excess_size);
+            if percent_distributable_block_size > Au::zero() {
+                for track_index in track_range.clone() {
+                    let row_deficit =
+                        get_percent_block_size_deficit(track_index, track_sizes[track_index]);
+                    if row_deficit > Au::zero() {
+                        let ratio =
+                            row_deficit.to_f32_px() / percent_block_size_deficit.to_f32_px();
+                        let size = percent_distributable_block_size.scale_by(ratio);
+                        track_sizes[track_index] += size;
+                        excess_size -= size;
```

Comment:
Can this make `excess_size` negative?

Reply:
I don't think so. We calculate the ratio via (row_deficit / total_deficit), but this ratio is multiplied by total_deficit.min(excess_size), so we always take a ratio of excess size.